beets/util/__init__.py

```diff
@@ -474,32 +475,45 @@ def copy(path, dest, replace=False):
                               traceback.format_exc())
 
 
-def move(path, dest, replace=False):
+def move(path, dest, replace=False, force=False):
     """Rename a file. `dest` may not be a directory. If `dest` already
     exists, raises an OSError unless `replace` is True. Has no effect if
-    `path` is the same as `dest`. If the paths are on different
-    filesystems (or the rename otherwise fails), a copy is attempted
-    instead, in which case metadata will *not* be preserved. Paths are
-    translated to system paths.
+    `path` is the same as `dest` unless `force` is True. If the paths are
+    on different filesystems (or the rename otherwise fails), a copy
+    is attempted instead, in which case metadata will *not* be preserved.
+    Paths are translated to system paths.
     """
-    if samefile(path, dest):
+    if os.path.isdir(path):
+        raise FilesystemError(u'source is directory', 'move', (path, dest))
+    if samefile(path, dest) and not force:
         return
     path = syspath(path)
     dest = syspath(dest)
-    if os.path.exists(dest) and not replace:
+    if os.path.exists(dest) and not replace and not force:
         raise FilesystemError(u'file exists', 'rename', (path, dest))
 
     # First, try renaming the file.
     try:
-        os.rename(path, dest)
+        os.replace(path, dest)
     except OSError:
         # Otherwise, copy and delete the original.
+        if os.path.isdir(dest):
+            dest = os.path.join(dest, os.path.basename(path))
+        tmp = tempfile.mktemp(suffix='.beets',
+                              prefix=as_string(b'.' + os.path.basename(dest)),
+                              dir=as_string(os.path.dirname(dest)))
+        tmp = syspath(tmp)
         try:
-            shutil.copyfile(path, dest)
+            shutil.copyfile(path, tmp)
+            os.replace(tmp, dest)
+            tmp = None
             os.remove(path)
```

Comment:
I haven't thought this through, but this seems like a relatively long `try` clause: Does this always do the best thing to recover at each of the possible points of failure?

Reply:
I not sure that you suggest here :(

Reply:
There's no direct suggestion, this is mostly a note to myself (and others!) to think about it in more depth. Maybe there's no issue, but at a first glance, this seemed like a critical section of the code to me.

Reply:
@wisp3rwind split it into a few try blocks increases code dramatically. I prefer to keep it simple.

Do you agree?